README.md
acd_cli
=======

**acd_cli** aims to provide a command line interface to Amazon Cloud Drive written in Python 3. 
It is currently in alpha stage.

##Features

 * local node caching
 * addressing of remote nodes via a pathname (e.g. `/Photos/kitten.jpg`)
 * plugin support

###File operations
 * tree or flat listing of files and folders
 * upload/download of single files and directories
 * folder creation
 * trashing/restoring
 * moving/renaming nodes

##Planned

 * shell completion for remote paths
 * "smart" folder syncing (maybe)
 * ... minor stuff

##Quick start

On the first start of the program (try ``./acd_cli.py sync``), you will have to complete the OAuth procedure.
A browser tab will open and you will be asked to log in or grant access for 'acd_cli_oa'.
Signing in or clicking on 'Continue' will download a JSON file named `oauth_data`,
which must be placed in the cache directory displayed on screen (e.g. `/home/<USER>/.cache/acd_cli`).

You may view the Appspot source code at https://tensile-runway-92512.appspot.com/src. 

##Usage

The following actions are built in

```
    sync (s)            refresh node list cache; necessary for many actions
    clear-cache (cc)    clear node cache [offline operation]
    tree (t)            print directory tree [offline operation]
    upload (ul)         file and directory upload to a remote destination
    overwrite (ov)      overwrite file A [remote] with content of file B [local]
    download (dl)       download a remote folder or file; will overwrite local files
    open (o)            open node
    create (c, mkdir)   create folder using an absolute path
    list-trash (lt)     list trashed nodes [offline operation]
    trash (rm)          move node to trash
    restore (re)        restore from trash
    children (ls)       list folder's children [offline operation]
    move (mv)           move node A into folder B
    rename (rn)         rename a node
    resolve (rs)        resolve a path to a node ID
    find (f)            find nodes by name [offline operation]
    add-child (ac)      add a node to a parent folder
    remove-child (rc)   remove a node from a parent folder
    usage (u)           show drive usage data
    quota (q)           show drive quota [raw JSON]
    metadata (m)        print a node's metadata [raw JSON]
```

Please run ``./acd_cli.py --help`` to get a current list of the available actions. 
You may also get a list of further arguments and their order of an action by calling ``./acd_cli.py [action] --help``.

You may provide most node arguments as a 22 character ID or a UNIX-style path. 
Trashed nodes' paths might not be able to be resolved correctly; use their ID instead.

When uploading/downloading large amounts of files, it is advisable to save the log messages to a file. 
This can be done by appending `2> >(tee acd.log >&2)` to the command.

Files can be excluded via optional parameter by file ending, e.g. `-xe bak`, 
or regular expression on the whole file name, e.g. `-xr "^thumbs\.db$"`. 
Both exclusion methods are case insensitive. 

When the script is done running, its exit status can be checked for flags. If no error occurs, the exit status 
will be 0. The flag values are: 
argument error -- 2,
failed file transfer -- 8,
upload timeout -- 16,
hash mismatch -- 32,
error creating folder -- 64.
If multiple errors  occur, their values will be compounded by a binary OR operation.

##Usage example

```
$ ./acd_cli.py sync
# Syncing... Done.
$ ./acd_cli.py tree
# [PHwiEv53QOKoGFGqYNl8pw] [A] /
$ ./acd_cli.py create /egg/
$ ./acd_cli.py create /egg/bacon/
$ ./acd_cli.py upload local/spam/ /egg/bacon/
# Current directory: local/spam/
# Current file: local/spam/sausage
# [##################################################] 100.00% of 20.0MiB
# Current file: local/spam/lobster
# [##################################################] 100.00% of 10.0MiB
# [...]
$ ./acd_cli.py tree
# [PHwiEv53QOKoGFGqYNl8pw] [A] /
# [         ...          ] [A] /egg/
# [         ...          ] [A] /egg/bacon/
# [         ...          ] [A] /egg/bacon/spam/
# [         ...          ] [A] /egg/bacon/spam/sausage
# [...]
```

##Known Issues

 * files larger than 10GiB may not be downloaded
 * the tree and recursive folder listings are prone to infinite recursion if folder loops exist

Feel free to use the bug tracker to add issues. You might find the `--verbose` and `--debug` options helpful. 

##Dependencies

 * dateutils
 * pycurl
 * requests >= 1.0.0
 * sqlalchemy

If you are using a Debian-based distribution, the necessary packages are ``python3-dateutil python3-pycurl python3-requests python3-sqlalchemy``.

##Contact

acd_cli@mail.com

##Changelog

### Version 0.1.3
 * OAuth now via Appspot; security profile no longer necessary
 * back-off algorithm for API requests implemented

### Version 0.1.2
new:
 * overwriting of files
 * recursive upload/download
 * hashing of downloaded files
 * clear-cache action

fixes:
 * remove-child accepted status code
 * fix for upload of files with Unicode characters
 
other:
 * changed database schema


HISTORY.rst
Ancient History
===============

0.1.3
-----
* plugin mechanism added
* OAuth now via Appspot; security profile no longer necessary
* back-off algorithm for API requests implemented

0.1.2
-----
new:
 * overwriting of files
 * recursive upload/download
 * hashing of downloaded files
 * clear-cache action

fixes:
 * remove-child accepted status code
 * fix for upload of files with Unicode characters

other:
 * changed database schema


cache/db.py
import os
import logging
from sqlalchemy import *
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, relationship
from datetime import datetime, timedelta

logger = logging.getLogger(__name__)

session = None
engine = None

Base = declarative_base()

DB_SCHEMA_VER = 1

parentage_table = Table('parentage', Base.metadata,
                        Column('parent', String(50), ForeignKey('folders.id'), primary_key=True),
                        Column('child', String(50), ForeignKey('nodes.id'), primary_key=True)
                        )


class Metadate(Base):
    """added in v1"""
    __tablename__ = 'metadata'

    key = Column(String(64), primary_key=True)
    value = Column(String)

    def __init__(self, key: str, value: str):
        self.key = key
        self.value = value


class Label(Base):
    """added in v1"""
    __tablename__ = 'labels'

    id = Column(String(50), ForeignKey('nodes.id'), primary_key=True)
    name = Column(String(256), primary_key=True)


# TODO: cycle safety for full_path()
class Node(Base):
    __tablename__ = 'nodes'

    # apparently 22 chars; max length is 22 for folders according to API doc ?!
    id = Column(String(50), primary_key=True, unique=True)
    type = Column(String(15))
    name = Column(String(256))
    description = Column(String(500))

    created = Column(DateTime)
    modified = Column(DateTime)
    updated = Column(DateTime)

    # "pending" status seems to be reserved for not yet finished uploads
    status = Column(Enum('AVAILABLE', 'TRASH', 'PURGED', 'PENDING'))

    __mapper_args__ = {
        'polymorphic_identity': 'node',
        'polymorphic_on': type
    }

    def __lt__(self, other):
        """ Compares this Node to another one on same path level. Sorts case-sensitive, Folder first. """
        if isinstance(self, Folder):
            if isinstance(other, File):
                return True
            return self.name < other.name
        if isinstance(other, Folder):
            return False
        return self.name < other.name

    def is_file(self) -> bool:
        return isinstance(self, File)

    def is_folder(self) -> bool:
        return isinstance(self, Folder)

    def is_available(self) -> bool:
        return self.status == 'AVAILABLE'

    def id_str(self) -> str:
        """short id string containing id, stat, name"""
        return '[{}] [{}] {}'.format(self.id, self.status[0], self.simple_name())

    def long_id_str(self, path: str=None) -> str:
        """long id string containing id, stat, path + name"""
        if path is None:
            path = self.containing_folder()
        return '[{}] [{}] {}{}'.format(self.id, self.status[0], path,
                                       ('' if not self.name else self.name)
                                       + ('/' if isinstance(self, Folder) else ''))

    def containing_folder(self) -> str:
        if len(self.parents) == 0:
            return ''
        return self.parents[0].full_path()

    parents = relationship('Folder', secondary=parentage_table,
                           primaryjoin=id == parentage_table.c.child,
                           backref='children'
                           )


class File(Node):
    __tablename__ = 'files'

    id = Column(String(50), ForeignKey('nodes.id'), primary_key=True, unique=True)
    md5 = Column(String(32))
    size = Column(BigInteger)

    __mapper_args__ = {
        'polymorphic_identity': 'file'
    }

    def __init__(self, id: str, name: str, created: datetime, modified: datetime, md5: str, size: int, status: Enum):
        self.id = id
        self.name = name
        self.created = created.replace(tzinfo=None)
        self.modified = modified.replace(tzinfo=None)
        self.md5 = md5
        self.size = size
        self.status = status

    def __eq__(self, other):
        if not isinstance(other, self.__class__):
            return False
        return (self.id == other.id and self.name == other.name
                and self.created - other.created == timedelta(0)
                and self.modified - other.modified == timedelta(0)
                and self.md5 == other.md5 and self.size == other.size
                and self.status == other.status)

    def __repr__(self):
        return 'File(%r, %r)' % (self.id, self.name)

    def simple_name(self) -> str:
        """file name"""
        return self.name

    def full_path(self) -> str:
        """absolute path of file (first containing folder chain)"""
        if len(self.parents) == 0:
            return self.name
        return self.parents[0].full_path() + self.name


class Folder(Node):
    __tablename__ = 'folders'

    id = Column(String(50), ForeignKey('nodes.id'), primary_key=True, unique=True)

    __mapper_args__ = {
        'polymorphic_identity': 'folder'
    }

    # noinspection PyShadowingBuiltins
    def __init__(self, id: str, name: str, created: datetime, modified: datetime, status: Enum):
        self.id = id
        self.name = name
        self.created = created.replace(tzinfo=None)
        self.modified = modified.replace(tzinfo=None)
        self.status = status

    def __eq__(self, other):
        if not isinstance(other, self.__class__):
            return false
        return (self.id == other.id
                and self.name == other.name
                and self.created == other.created
                and self.modified == other.modified
                and self.status == other.status)

    def __repr__(self):
        return 'Folder(%r, %r)' % (self.id, self.name)

    def simple_name(self) -> str:
        return (self.name if self.name else '') + '/'

    # path of first occurrence
    def full_path(self) -> str:
        if len(self.parents) == 0:
            return '/'
        return self.parents[0].full_path() + self.simple_name()

    def get_child(self, name: str) -> Node:
        """ Gets non-trashed child by name. """
        for child in self.children:
            if child.name == name and child.status != 'TRASH':
                return child
        return


"""End of 'schema'"""


def init(path=''):
    logger.info('Initializing cache with path "%s".' % os.path.realpath(path))
    db_path = os.path.join(path, 'nodes.db')

    global session
    global engine
    engine = create_engine('sqlite:///%s' % db_path)

    uninitialized = not os.path.exists(db_path)
    if not uninitialized:
        uninitialized = not engine.has_table(Node.__tablename__)
    if uninitialized:
        r = engine.execute('PRAGMA user_version = %i;' % DB_SCHEMA_VER)
        r.close()

    logger.info('Cache %sconsidered uninitialized.' % ('' if uninitialized else 'not '))

    Base.metadata.create_all(engine)
    _session = sessionmaker(bind=engine)
    session = _session()

    if uninitialized:
        return

    import cache.sync

    r = engine.execute('PRAGMA user_version;')
    ver = r.first()[0]
    r.close()

    logger.info('DB schema version is %s.' % ver)

    if DB_SCHEMA_VER > ver:
        _migrate(ver)

    oldest = cache.sync.max_age()
    if oldest:
        logger.info('Oldest node info is %f days old.' % oldest)
        if oldest > 30:
            logger.warning('Cache is outdated. Please perform a full sync.')


def drop_all():
    Base.metadata.drop_all(engine)
    logger.info('Dropped all tables.')


def _migrate(schema: int):
    """ Migrate database to highest schema
    :param schema: current (cache file) schema version to upgrade from
    """

    migrations = [_0_to_1]

    conn = engine.connect()
    trans = conn.begin()
    try:
        for update_ in migrations[schema:]:
            logger.warning('Updating db schema from %i to %i.' % (schema, schema + 1))
            update_(conn)
            schema += 1
        trans.commit()
    except:
        trans.rollback()
        raise

    conn.close()


def _0_to_1(conn):
    conn.execute('ALTER TABLE nodes ADD updated DATETIME;')
    conn.execute('ALTER TABLE nodes ADD description VARCHAR(500);')
    conn.execute('PRAGMA user_version = 1;')